File: API/BusTicket.Core/Models/User/UserUpsertModel.cs
namespace BusTicket.Core.Models
{
    public class UserUpsertModel : BaseUpsertModel
    {
        public bool IsActive { get; set; }
        public bool IsFirstLogin { get; set; }
        public bool VerificationSent { get; set; }
        public string FirstName { get; set; } = default!;
        public string LastName { get; set; } = default!;
        public string UserName { get; set; } = default!;
        public string NormalizedUserName { get; set; } = default!;
        public string Email { get; set; } = default!;
        public string NormalizedEmail { get; set; } = default!;
        public bool EmailConfirmed { get; set; }
        public string PhoneNumber { get; set; } = default!;
        public bool PhoneNumberConfirmed { get; set; }
        public int AccessFailedCount { get; set; }
        public DateTime? BirthDate { get; set; }
        public Gender? Gender { get; set; }
        public string? ProfilePhoto { get; set; }
        public string? ProfilePhotoThumbnail { get; set; }
        public string Address { get; set; } = default!;
        public bool IsAdministrator { get; set; }
        public bool IsClient { get; set; }
        public bool EnableNotificationEmail { get; set; }

        public List<UserRole> UserRoles { get; set; } = new List<UserRole>();

        //password
        public string? OldPassword { get; set; }
        public string? NewPassword { get; set; }
    }
}

File: UI/bus_ticket_admin/lib/models/user_upsert_model.dart
// user_profile.dart
import 'package:bus_ticket_admin/models/user_role.dart';

class UserUpsertModel {
  int? id;
  String firstName;
  String lastName;
  String userName;
  String email;
  String phoneNumber;
  DateTime? birthDate;
  int? gender;
  String? profilePhoto;
  String address;
  List<UserRoleModel>? userRoles;

  UserUpsertModel({
    this.id,
    required this.firstName,
    required this.lastName,
    required this.userName,
    required this.email,
    required this.phoneNumber,
    this.birthDate,
    this.gender,
    this.profilePhoto,
    required this.address,
    this.userRoles,
  });

  factory UserUpsertModel.fromJson(Map<String, dynamic> json) {
    return UserUpsertModel(
      id: json['id'],
      firstName: json['firstName'],
      lastName: json['lastName'],
      userName: json['userName'],
      email: json['email'],
      phoneNumber: json['phoneNumber'],
      birthDate: json['birthDate'] != null ? DateTime.parse(json['birthDate']) : null,
      gender: json['gender'],
      profilePhoto: json['profilePhoto'],
      address: json['address'],
      userRoles: json['userRoles'] != null ? (json['userRoles'] as List).map((e) => UserRoleModel.fromJson(e)).toList() : null,
      //role: RoleLevel.values.elementAt(json['gender']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'firstName': firstName,
      'lastName': lastName,
      'userName': userName,
      'email': email,
      'phoneNumber': phoneNumber,
      'birthDate': birthDate?.toIso8601String(),
      'gender': gender,
      'profilePhoto': profilePhoto,
      'address': address,
      'userRoles': userRoles
    };
  }
}